### PR TITLE
Revise Query Check

### PR DESCRIPTION
**Pull request by rdvelazquez**
This is a work in progress PR to addresses #102. I'm posting this now to:
1. Let everyone know that I'm working on this and avoid duplicating work.
2. Get some advice or guidance.

Once complete this PR will add an additional check so that users can’t submit queries if the query doesn’t have at least 20 positive and 20 negative samples. See #102 for the reasoning. I also revised the checks so that the classifier is validated before asking for an email address (hence the rearranging of things and the addition of the new function `validateClassifier`.

This is my first time working with the frontend so I’m not completely familiar with how things are set up. I left dummy variables in line 41 of `QueryBuilder.js` for the number of positive samples and number of negative samples in the query. I’m not sure the best way to get the number of positives and number of negatives… the numbers shown on cognoma.org under DISEASE TYPE SET are set by `.setTotalFor` in the following two files:
 - [querySetDiseaseTpye.tpl.html](https://github.com/cognoma/frontend/blob/develop/app/js/components/queryBuilder/querySets/querySetDiseaseType/querySetDiseaseType.tpl.html), specifically line 24 `{{$ctrl.setTotalFor('positives')}}` and line 28 `{{$ctrl.setTotalFor('negatives')}}` and
 - [querySetDiseaseType.component.js]( https://github.com/cognoma/frontend/blob/develop/app/js/components/queryBuilder/querySets/querySetDiseaseType/querySetDiseaseType.component.js), specifically line 25 `vm.setTotalFor = setParam=>_.reduce(_.pluck(this.diseaseSet, setParam), (a,b)=>a+b);`

@dcgoss and/or @bdolly... I'll look into this a little more and update it if I find a working solution but any advice or guidance would be much appreciated.

### CODE REVIEW

**Comment by rdvelazquez** [comment]
This is now working on my local branch and is ready for a review. @bdolly and I met last night and he was able to show me how to get it working. Thanks!

@dhimmel do you want to get this merged before tomorrow night?

**Comment by dhimmel** [comment]
> @dhimmel do you want to get this merged before tomorrow night?

That would be great! Thanks @rdvelazquez for dabbling in the frontend.

I'm not a great reviewer for this. @bdolly did you look at this and is it good to go?

**Comment by rdvelazquez** [comment]
@bdolly this is ready for your review/merge.

**Comment by dhimmel** [comment]
In case @bdolly is busy, I'm tagging @dongbohu (a full stack developer that's part of the Greene Lab and CCDL). Would love the get this reviewed and merged.

File: app/js/services/QueryBuilderService.js [hunk 1]
```diff
@@ -36,90 +36,105 @@ function QueryBuilderService( $resource, $http, AppSettings, NotificationService
...
           });
-        }
+        });
+
       }
 
-      if (user.email === null || user.email === undefined || IGNORE_EXISTING_EMAIL) {
-        // update email
-        let email = prompt("Enter your email address to receive your classifier:")
+      function getEmail(){
```
Comment (dongbohu) [inline app/js/services/QueryBuilderService.js]: This function not only gets email but also calls `submitClassifier()`. You may want to rename it to something like `doSubmission`.

File: app/js/services/QueryBuilderService.js [hunk 2]
```diff
@@ -36,90 +36,105 @@ function QueryBuilderService( $resource, $http, AppSettings, NotificationService
...
+      function validateClassifier() {
+        let number_positives = getTotalsFor('positives')
+        let number_negatives = getTotalsFor('negatives')
+        // 20 positive samples and 20 negative samples
+        if(number_positives > 20 && number_negatives > 20){
+          return true
+        } else {
+          NotificationService.notify({
+            type:    'error',
+            message: 'You must select at least 1 disease and 1 gene; and the query must have at least 20 positive and 20 negative samples'
```
[the remark this answers is not in the document]
Comment (rdvelazquez) [reply app/js/services/QueryBuilderService.js]: >Both number_positives and number_negatives are calculated based on diseases only. What if both number_positives and number_negatives are larger than 20, but genes.length is still 0?

There will always be 0 positives and 0 negatives if genes.length is 0 or diseases.length is 0 so it seemed redundant to include the check for genes.length or diseases.length in the code. but I do think keeping "You must select at least 1 disease and 1 gene" in the message/prompt may be useful for some users.

[the remark this answers is not in the document]
Comment (rdvelazquez) [reply app/js/services/QueryBuilderService.js]: >Besides, the code on line 46 seems to require more than (aka. >) 20 positives and negatives, NOT "at least" (aka. >=).

Good catch! I've revised the code to greater than or equal to.

File: app/js/services/QueryBuilderService.js (hunk 1 above)
Comment (rdvelazquez) [reply app/js/services/QueryBuilderService.js]: I renamed it to `executeSubmission`. It's not a great name but much better than `getEmail`.

**Comment by rdvelazquez** [comment]
Thank you for your review @dongbohu! Very helpful comments!

>Obviously there are many coding style issues in the whole code base (not only this PR). I pointed out a few but eventually gave up. I strongly recommend you guys set up some tools such as codeclimate to make the code more readable.

I agree that the coding style (as well as structural layout) does not lend itself to readability. I think something like code climate could be a good addition moving forward.

File: app/js/services/QueryBuilderService.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-function QueryBuilderService( $resource, $http, AppSettings, NotificationService) {
+function QueryBuilderService( $resource, $http, AppSettings, NotificationService, _) {
```
[the remark this answers is not in the document]
Comment (rdvelazquez) [reply app/js/services/QueryBuilderService.js]: I removed the space after `(`

I think it's common practice to just use `_` for [underscore.js](http://underscorejs.org/) but I can change it to something else if you prefer.

File: app/js/services/QueryBuilderService.js [hunk 4]
```diff
@@ -36,90 +36,105 @@ function QueryBuilderService( $resource, $http, AppSettings, NotificationService
...
               });
 
+      // Helper function to get the total number of positive or negative samples in a query
+      let getTotalsFor = setParam=>_.reduce(_.pluck(diseases, setParam), (a,b)=>a+b)
+
+      function validateClassifier() {
+        let number_positives = getTotalsFor('positives')
+        let number_negatives = getTotalsFor('negatives')
+        // 20 positive samples and 20 negative samples
+        if(number_positives > 20 && number_negatives > 20){
```
[the remark this answers is not in the document]
Comment (rdvelazquez) [reply app/js/services/QueryBuilderService.js]: Addressed.

File: app/js/services/QueryBuilderService.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rdvelazquez) [reply app/js/services/QueryBuilderService.js]: > Style suggestion: This line is so long. You may want to break it into two.

Good suggestion. I'll tried to keep local consistence by not wrapping long lines (as none of the other long lines are wrapped) but I may look at reformatting some of this code at a later date and that would be something to include in the reformatting.

**Comment by bdolly** [comment]
>Obviously there are many coding style issues in the whole code base (not only this PR). I pointed out a few but eventually gave up. I strongly recommend you guys set up some tools such as codeclimate to make the code more readable.

Totally agree and I think something like codeclimate is a great idea. However, seeing as how the codebase that exists was intended as MVP only and to be refactored into a more updated js framework (React, Angular, Vue)  not sure that tooling would be worth the effort and clean-up at this time.

**Comment by bdolly** [comment]
@dcgoss @dongbohu All looks good to me, my only suggestion would be to make the notification bar stay open a bit longer so the user has more time to read it. 

If both of you give this the thumbs up, I'll be happy to merge.

**Comment by dhimmel** [comment]
> until the testing suite and CI is ironed out and fully implemented

Automatic deployment would be really convenient. If I understand correctly, this repo will automatically deploy to AWS upon a successful build?

What is failing with the tests? Perhaps we should just skip the tests until they're fully implemented? The test library probably has some option to allow failure of a test.

**Pull request merged**